// @allin/wap-share/shareAll.js
/**
 * @component-name:
 * @desc: 调用方法：shareAll(obj,weChat,trigger);
 *  (【obj必填，可为{}】)，weChat(必填) [true只做微信分享，false分享微信及微博，空间], trigger（选填[$('#id')或$('.class')])，
 * @example:
 *          shareAll({
                weixinConfigUrl:'',                                             //【可选】获取微信配置接口，【默认：唯医的】
                // weixinMask:"<section class="ev-weixinMask"></section>",         //【可选】可配置微信浏览器中分享遮罩模板 ★★注意：在配置遮罩模板时必须带 ev-weixinMask 类【默认的类样式在唯医base.css中】
                shareTemplate:'<section class="ev-shareTemplate"></section>'，  //【可选】可以配置分享模板   ★★注意：分享模板顶级元素加类名 ev-shareTemplate 并且加“on”类代表显示，去除‘on’隐藏
                title:'11113',                                                  //【可选】分享标题
                url:"http://www.allinmd.cn",                                    //【可选】不传默认取当前地址
                pic:"",                                                         //【可选】分享图片
                trendUrl:"",                                                    //【可选】分享动态的请求连接  //不需要动态分享就不传
                noPJ:0,                                                         //【可选】0：分享动态给后台的参数需要转为paramJson ,1:不需要
                data:{},                                                        //【可选】分享动态传给后台的参数
                callback:function(){},                                          //【可选】分享动态后成功后的回调函数
                wxTitle:'wxxxx',                                                //【可选】微信分享好友标题【默认取title】
                wxDesc:'',                                                      //【可选】微信分享好友描述
                timeLineTitle:'timelIne',                                       //【可选】分享到朋友圈标题【默认取title】
                sinaTitle:'sina description',                                   //【可选】分享到新浪的话术【默认取title】
                desc:"",                                                        //
                qzoneTitle:'',                                                  //【可选】qq空间title【默认取title】
                summary:"this is summary",                                      //【可选】qq空间的描述
                noSelfInit:  true or false,                                     //【可选】是否自执行【默认false】，使用场景列表类的分享点击时再触发 Weixin_share
                noWXShare:false,                                                //【可选】是否禁止微信分享话术【默认false】
                qShareLog:function(x){                                          //【可选】分享日志：新浪，空间点击按钮即执行
                    if(x=='sina'){
                        console.log('sina');
                    }else{
                        console.log('qzone');
                    }
                },
                trendCallBack:function(){},                                     //唯医之外的其他站点需要做朋友圈，动态分享时自定义函数
                fnMessageSuc:function(){},                                      //【可选】分享好友成功回调
                fnTimelineSuc:function(){}，                                      //【可选】分享朋友圈成功回调
                triggerEventFn:function(){}                                        //点击trigger时事件函数（埋点）
            },false,$('.share'));
 * @depend: jquery, jweixin-1.0.0.js[唯医分享动态朋友圈额外需要引用comm.func.js,module.user.js,"/js/comm/comm-customer.js","/js/comm/comm-data.js","/js/comm/comm_service.js"]
 * @warn:   ★★注意1：在配置遮罩模板时必须带 ev-weixinMask 类
 *          ★★注意2：在非微信浏览器中，分享模板顶级元素加类名 ev-shareTemplate 并且加“on”类代表显示，去除‘on’隐藏
 *          ★★注意3：在非微信浏览器中，点击分享到空间，新浪和动态的每个项目有个属性 “ref” ，且类名为ev-shareTemplateItem，取消分享的按钮id #EV-cancelShare
 *          ★★注意4: 不需要再额外给trigger加click事件，且trigger为$(dom)元素
 * @udpate: 1 :jweixin-1.0.0.js 内置到些函数上方，无需再单独引用
 *          2 :非微信浏览器不再请求jsconfig参数
 *          3: 新增经纬度获取 （2017-7-19）
 * @date: modified on 2017/03/07
 * @author: sunhaibin
 */
import wx from 'weixin-jsapi';
export  function shareAll(obj, weChat, trigger,onlyAllinCircle) {
	var shareDef = {
		title: document.title,
		url: window.location.href,
		pic: 'https://m.allinmd.cn/image/allin_logo.png',
		summary: "汇集海量手术视频、医学文献、病例讨论、会议回放资源。医生专业互动社区。", //qzone描述
		sinaTitle: '',
		sinaDesc: '',
		qzoneTitle: '',
		qqTitle: '',
		wxTitle: "", //微信分享标题
		wxDesc: "", //微信分享描述
		timeLineTitle: "",
		desc: "汇集海量手术视频、医学文献、病例讨论、会议回放资源。医生专业互动社区。",
		closeDialog: function() {
			if ($('.ev-weixinMask').length) {
				$('.ev-weixinMask').remove();
				$("body").unbind('touchmove');
			}
		},
		// 分享好友成功
		fnMessageSuc: function() {
		
		},
		// 分享朋友圈成功
		fnTimelineSuc: function() {
		
		},
		//分享新浪及空间成功（shareToName为分享渠道，if(shareToName=='sina'){fn();})）
		qShareLog: function(shareToName) {
		
		}
	};
	var commShareData = undefined,jsConfigData;
	var getStrLen = function(str, len) {
		var strlen = 0,
			s = "";
		for (var i = 0; i < str.length; i++) {
			s = s + str.charAt(i);
			if (str.charCodeAt(i) > 128) {
				strlen = strlen + 2;
				if (strlen >= len) {
					return s.substring(0, s.length - 1) + "...";
				}
			} else {
				strlen = strlen + 1;
				if (strlen >= len) {
					return s.substring(0, s.length - 2) + "...";
				}
			}
		}
		return s;
	};
	var Weixin_share = function(obj) {
		var controller = {
			init: function(obj,autoShare) {
				var t = this;
				t.options = {};
				t.options = $.extend(shareDef, obj);
				if(autoShare && t.options.triggerRequest){
					if(!commShareData){
						commShareData = t.options.triggerRequest();
					}
					t.options = $.extend(t.options, commShareData);
				}
				t.options.desc = getStrLen(t.options.desc,50);
				t.options.summary = getStrLen(t.options.summary,50);
				t.options.pic = /^http/.test(t.options.pic)?t.options.pic:"https:"+t.options.pic;   //防止没有http或https协议导航图片找不到
				//if (obj.triggerCallback) {
				//    t.options = $.extend(t.options,obj.triggerCallback());
				//}
				t.getWxConfig();
			},
			getWxConfig: function() {
				var t = this;
				var data;
				var jsApiList = ['onMenuShareTimeline','onMenuShareAppMessage','onMenuShareQQ','onMenuShareQZone'];
				if(obj.noWXShare){
					jsApiList = ['onMenuShareTimeline','onMenuShareAppMessage','onMenuShareQQ','onMenuShareQZone','hideOptionMenu'];
				}
				/*
				 *  weixinConfigUrl 请求微信配置接口，各网站各自配置
				 * */
				//setTimeout(function(){
				if(!jsConfigData){
					$.ajax({
						url: obj.weixinConfigUrl?obj.weixinConfigUrl:"/mcall/wx/api/getJSConfig/",
						type: "post",
						data: {
							paramJson: JSON.stringify({
								url: location.protocol+'//'+location.host+location.pathname+(location.search!=''?location.search:(location.href.indexOf('?')>-1?'?':''))	//TODO 在ios中url中带#参数请求返回的signature无效
							})
						},
						dataType: 'json',
						async: true,
						success: function(res) {
							data = res.responseObject.responseData;
							jsConfigData = data;
							wx.config({
								debug: false,
								appId: data.appId,
								timestamp: data.timestamp,
								nonceStr: data.noncestr,
								signature: data.signature,
								jsApiList:jsApiList
							});
							if(obj.noWXShare){
								wx.hideOptionMenu();
							}else{
								setTimeout(function() {
									t.shareContent();
								}, 500);
							}
						}
					});
				}
				
				//},500);
				
			},
			shareContent: function() {
				var t = this;
				wx.ready(function() {
					wx.onMenuShareAppMessage({
						title: t.options.wxTitle != '' ? t.options.wxTitle : t.options.title,
						desc: t.options.wxDesc ? t.options.wxDesc : t.options.desc,
						link: t.options.url,
						imgUrl: t.options.pic,
						success: function() {
							t.options.closeDialog();
							if (t.options.fnMessageSuc) {
								t.options.fnMessageSuc(t.options);
							}
							if(window.comm&&window.comm.creatEvent){
								comm.creatEvent({
									triggerType:'分享',
									trigger_name:"分享好友",
									actionId:70
								});
							}
						},
						cancel: function() {
							t.options.closeDialog();
						}
					});
					wx.onMenuShareTimeline({
						title: t.options.timeLineTitle === "" ? t.options.title : t.options.timeLineTitle,
						desc: t.options.wxDesc ? t.options.wxDesc : t.options.desc,
						link: t.options.url,
						imgUrl: t.options.pic,
						success: function() {
							t.options.closeDialog();
							if (t.options.fnTimelineSuc) {
								t.options.fnTimelineSuc(t.options);
							}
							if(window.comm&&window.comm.creatEvent){
								comm.creatEvent({
									triggerType:'分享',
									trigger_name:"分享朋友圈",
									actionId:70
								});
							}
						},
						cancel: function() {
							t.options.closeDialog();
						}
					});
					wx.onMenuShareQQ({
						title: t.options.qqTitle === "" ? t.options.title : t.options.qqTitle,
						desc: t.options.wxDesc ? t.options.wxDesc : t.options.desc,
						link: t.options.url,
						imgUrl: t.options.pic,
						success: function() {
							t.options.closeDialog();
							if(window.comm&&window.comm.creatEvent){
								comm.creatEvent({
									triggerType:'分享',
									trigger_name:"分享QQ",
									actionId:70
								});
							}
						},
						cancel: function() {
							t.options.closeDialog();
						}
					});
					wx.onMenuShareQZone({
						title: t.options.qzoneTitle === "" ? t.options.qzoneTitle : t.options.qzoneTitle,
						desc: t.options.desc,
						link: t.options.url,
						imgUrl: t.options.pic,
						success: function () {
							t.options.closeDialog();
							if(window.comm&&window.comm.creatEvent){
								comm.creatEvent({
									triggerType:'分享',
									trigger_name:"分享QQ空间",
									actionId:70
								});
							}
						},
						cancel: function () {
							t.options.closeDialog();
						}
					});
					wx.error(function (res) {
						console.log(res);
						// config信息验证失败会执行error函数，如签名过期导致验证失败，具体错误信息可以打开config的debug模式查看，也可以在返回的res参数中查看，对于SPA可以在这里更新签名。
					});
				})
			}
		};
		if(isWeixin ()){
			controller.init(obj,'autoShare');
		}
	};
	var mobileShare = function(obj) {
		var _controller = {
			default: {
				qzone_m: "http://qzs.qzone.qq.com/open/connect/widget/mobile/qzshare/index.html?",
				sina_m: "http://v.t.sina.com.cn/share/share.php?&appkey=&"
				//sina_m: "http://service.weibo.com/share/mobile.php?&appkey=&"
				
			},
			init: function(obj) {
				var t = this;
				var opt = {};
				opt = $.extend(shareDef, obj);
				opt.desc = getStrLen(opt.desc,50);
				opt.summary = getStrLen(opt.summary,50);
				opt.pic = /^http/.test(opt.pic)?opt.pic:location.protocol+opt.pic;
				if(!opt.noSelfInit && isWeixin ()){//让其自执行分享话术 前提是在微信浏览器中
					Weixin_share(opt);
				}
				
				//微信分享
				t.shareAction(opt);
			},
			shareAction: function(opt) {
				var t = this;
				
				trigger.off('click').on('click', function() {
					if(opt.triggerRequest){//点击后请求分享话术
						if(!commShareData){
							commShareData= opt.triggerRequest();
						}
						if($.isEmptyObject(commShareData)){
							return false;
						}
						opt = $.extend({},opt, commShareData);
					}
					if (opt.triggerCallback) {
						opt = $.extend(opt, opt.triggerCallback($(this)));
					}
					var sl = $(this);
					if(opt.triggerEventFn){
						opt.triggerEventFn(sl);
					}
					//点击分享埋点
					if(window.comm&&window.comm.creatEvent){
						comm.creatEvent({
							triggerType:'分享',
							triggerName:'分享（呼出）',
							keyword:"分享（呼出）",
							actionId:40,
							locationId:(function(){
								if(sl.hasClass('ev-share')){
									if(sl.parents('.al-commentItem').length){
										if($('.icon-shareWhite').length){
											return sl.parents('.al-commentItem').index()+2;
										}else{
											return sl.parents('.al-commentItem').index()+1;
										}
									}else{
										return 1;
									}
								}else{
									return 1;
								}
								
							})(),
							refId:(function(){
								if(window.location.href.indexOf('/html/')>-1){      //终端页加refId
									var _ind = window.location.href.lastIndexOf('/')+1;
									return window.location.href.substring(_ind,_ind+13);
								}else{
									var _tId = comm.getpara().tId;      //标签专题页加tId
									if(_tId){
										return _tId;
									}else{
										return "";
									}
								}
							})()
						});
					}
					
					
					//if(opt.noSelfInit && isWeixin () ){
					//    Weixin_share(opt);
					//}
					if (isWeixin()) {//微信浏览器中
						/*
						 * 可配置微信浏览器中分享遮罩模板 ★★注意：在配置遮罩模板时必须带 ev-weixinMask 类
						 * */
						var shareHtml = opt.weixinMask? opt.weixinMask:'<div class="dk-result-pointer-bg ev-weixinMask"></div><div class="dk-result-pointer ev-weixinMask" ></div>';
						if($('.ev-weixinMask').length==0){
							$("body").append(shareHtml);
						}
						
						//禁止滚动
						$("body").bind('touchmove', function(e) {
							e.preventDefault();
						});
						
						$(".ev-weixinMask").on("click", function() {
							$(".ev-weixinMask").remove();
							$('body').unbind('touchmove');
							return false;
						});
						Weixin_share(opt);
					} else { //非微信浏览器
						if ($('.ev-shareTemplate').length > 0) { //已加载过弹层
							$('.ev-shareTemplate').addClass('on');
							$("body").on("touchmove", function(e) {
								e.preventDefault();
							});
						} else {
							/*@fixed
							 *@desc 样式重构
							 *      增加缓入缓出动效
							 *@fixed by qiangkailiang on 2016/08/18
							 */
							var trendImg = "";
							
							if (typeof(shareDef.trendUrl) !== 'undefined' && shareDef.trendUrl.length !== 0&&shareDef.trendUrl!="") {
								trendImg = '<img src="//img50.allinmd.cn/pages/index/share_allin.png" _mce_src="//img50.allinmd.cn/pages/index/share_allin.png" alt="">';
							} else {
								trendImg = '<img src="//img50.allinmd.cn/pages/index/share_allin_no.png" _mce_src="//img50.allinmd.cn/pages/index/share_allin_no.png" alt="">';
							}
							
							var notWeixinDom = '<section class="al-bottomSelectorListMask al-bottomSharePart ev-shareTemplate">' +
								'        <section class="al-bottomSelectorListBox">' +
								'            <section class="al-bottomSelectorList">' +
								'                <figure class="al-bottomSelectorItem al-bottomShareListBox">' +
								'                    <h2>分享给好友 共同讨论</h2>' +
								'                    <ul>' +
								'                        <li class="al-bottomShareItem ev-shareTemplateItem" ref="trend" style="display: none;">' +
								'                            <figure class="al-bottomShareImg">' +
								trendImg +
								'                            </figure>' +
								'                            <figcaption>唯医朋友圈</figcaption>' +
								'                        </li>' +
								'                        <li class="al-bottomShareItem share_qzone ev-shareTemplateItem" ref="qzone">' +
								'                            <figure class="al-bottomShareImg">' +
								'                                <img src="//img50.allinmd.cn/pages/index/share_kongjian.png" _mce_src="//img50.allinmd.cn/pages/index/share_kongjian.png" alt="">' +
								'                            </figure>' +
								'                            <figcaption>QQ空间</figcaption>' +
								'                        </li>' +
								'                        <li class="al-bottomShareItem share_sina ev-shareTemplateItem" ref="sina">' +
								'                            <figure class="al-bottomShareImg">' +
								'                                <img src="//img50.allinmd.cn/pages/index/share_weibo.png" alt="">' +
								'                            </figure>' +
								'                            <figcaption>微博</figcaption>' +
								'                        </li>' +
								'                    </ul>' +
								'                </figure>' +
								'            </section>' +
								'            <section class="al-bottomSelectorList mgt">' +
								'                <figure class="al-bottomSelectorItem" id="EV-cancelShare"><a href="javascript:void(0)">取消</a></figure>' +
								'            </section>' +
								'        </section>' +
								'    </section>';
							
							var notWeixinDom2 = '<section class="al-bottomSelectorListMask al-bottomSharePart ev-shareTemplate">' +
								'        <section class="al-bottomSelectorListBox">' +
								'            <section class="al-bottomSelectorList">' +
								'                <figure class="al-bottomSelectorItem al-bottomShareListBox">' +
								'                    <h2>分享给好友 共同讨论</h2>' +
								'                    <ul>' +
								'                        <li class="al-bottomShareItem ev-shareTemplateItem" ref="trend" style="display: none">' +
								'                            <figure class="al-bottomShareImg">' +
								trendImg +
								'                            </figure>' +
								'                            <figcaption>唯医朋友圈</figcaption>' +
								'                        </li>' +
								'                    </ul>' +
								'                </figure>' +
								'            </section>' +
								'            <section class="al-bottomSelectorList mgt">' +
								'                <figure class="al-bottomSelectorItem" id="EV-cancelShare"><a href="javascript:void(0)">取消</a></figure>' +
								'            </section>' +
								'        </section>' +
								'    </section>';
							/*
							 * 可以配置分享模板   ★★注意：分享模板顶级元素加类名 ev-shareTemplate 并且加“on”类代表显示，去除‘on’隐藏
							 * */
							if(opt.shareTemplate){
								$('body').append(opt.shareTemplate);
								
								$("body").on("touchmove", function(e) { //禁止滚动
									e.preventDefault();
								});
							}else{
								if(onlyAllinCircle){
									$('body').append(notWeixinDom2);
								}else{
									$('body').append(notWeixinDom);
								}
							}
							setTimeout(function() {
								$(".ev-shareTemplate").addClass('on');
								$("body").on("touchmove", function(e) {//禁止滚动
									e.preventDefault();
								});
							}, 200);
							
						}
						//★★注意：在非微信浏览器中，点击分享到空间，新浪和动态的每个项目有个属性 “ref” ，且类名为ev-shareTemplateItem
						$('.ev-shareTemplateItem').off("click").on('click', function() {
							var shareToName = $(this).attr('ref');
							/*@fixed
							 *@desc 增加动态的分享
							 *@fixed by lichunhui on 2016/08/23
							 */
							if (shareToName == "trend") {
								if (opt.trendUrl) { //需要分享动态
									if(opt.trendCallBack&&typeof opt.trendCallBack == 'function'){
										opt.trendCallBack();
									}else {
										user.privExecute({
											operateType: 'auth', //'login','auth','conference'
											callback: function () {
												var data = opt.noPJ ? opt.data : {
													paramJson: JSON.stringify(opt.data)
												};
												$.ajax({
													type: "post",
													url: opt.trendUrl,
													data: data,
													dataType: "json",
													success: function (rep) {
														$('.al-bottomSharePart').removeClass('on');
														$("body").css("overflow", "visible")
														if (rep && rep.responseObject.responseStatus) {
															opt.callback && opt.callback();
															popup("分享到唯医朋友圈成功！");
														} else {
															popup("分享到唯医朋友圈失败！");
														}
														$('body').unbind('touchmove');
													},
													error: function () {
														popup("分享到唯医朋友圈失败！");
														$('body').unbind('touchmove');
													}
												});
												comm.creatEvent({
													triggerType: '分享',
													triggerName: "分享动态",
													keyword: "分享动态",
													actionId: 70
												});
											}
										});
									}
								}
								
							} else {
								var newHref = t.default[shareToName + '_m'];
								var picture = '&pics=http:',
									shareTitle;
								if (shareToName == 'sina') {
									picture = "&pic=http:";
									shareTitle = opt.sinaTitle != '' ? opt.sinaTitle.split(':http')[0] : opt.title;
								} else if (shareToName == 'qzone') {
									shareTitle = opt.qzoneTitle != '' ? opt.qzoneTitle : opt.title;
								}
								if (opt.pic.search(':') >= 0) {
									opt.pic = opt.pic.split(':')[1];
								}
								
								opt.qShareLog && opt.qShareLog(shareToName,opt);
								if(window.comm&&comm.creatEvent) {
									comm.creatEvent({
										triggerType: '分享',
										triggerName:"分享"+(shareToName == 'qzone'?"QQ空间":"新浪微博"),
										keyword: "分享"+(shareToName == 'qzone'?"QQ空间":"新浪微博"),
										actionId: 70
									});
								}
								window.open(newHref + 'url=' + encodeURIComponent(opt.url) + '&title=' + encodeURIComponent(shareTitle) + picture + encodeURIComponent(opt.pic) + '&summary=' + encodeURIComponent(opt.summary) + (shareToName == 'sina' ? '&desc=' + encodeURIComponent(opt.desc) : ""), '_blank');
							}
							$('body').unbind('touchmove');
							return false;
						});
						$('#EV-cancelShare').on('click', function() {
							$('.ev-shareTemplate').removeClass('on');
							$("body").unbind("touchmove");
							return false;
						})
					}
					return false;
				})
			}
		};
		_controller.init(obj);
	};
	function isWeixin () { //是否微信浏览器
		var nav = navigator.userAgent;
		if (/MicroMessenger/.test(nav)) {
			return true;
		} else {
			return false;
		}
	}
	if (weChat == true) {
		Weixin_share(obj);
	} else {
		mobileShare(obj, false, trigger);
	}
}
